apps/sms/js/drafts.js

```diff
@@ -32,183 +26,107 @@
     get size() {
       return draftIndex.size;
     },
+
     /**
-     * add
-     *
-     * Push a Draft object or an object that has
-     * all the properties of a Draft instance
-     * to the Drafts collection. If the object
-     * isn't an instance of Draft, initialize
-     * a new Draft.
-     *
-     * @param  {Object}  draft  Draft-like object.
-     *
-     * @return {Drafts} return the Drafts object.
+     * Pushes a Draft instance to the Drafts in-memory collection and commits
+     * it to the persistent storage.
+     * @param  {Draft} draft Draft instance.
+     * @returns {Drafts} Returns the Drafts object.
      */
     add: function(draft) {
-      var threadId;
-      var thread;
-      var stored;
-
-      if (draft) {
-        if (!(draft instanceof Draft)) {
-          draft = new Draft(draft);
-        }
+      if (!draft || !(draft instanceof Draft)) {
+        throw new Error('Draft should be valid Draft instance.');
+      }
 
-        threadId = draft.threadId || null;
-        thread = draftIndex.get(threadId) || [];
-        stored = thread[thread.length - 1];
+      // Remove existing draft.
+      this.delete(draft);
 
-        // If there is an existing draft for this
-        // threadId, delete it.
-        // This should be replaced by a general
-        // replacement method.
-        if (threadId !== null && thread.length) {
-          this.delete(stored);
-        }
+      var drafts = draftIndex.get(draft.threadId) || [];
+      drafts.push(draft);
+      draftIndex.set(draft.threadId, drafts);
 
-        // If there is an existing threadless draft
-        // with the same id, delete it.
-        // This should be replaced by a general
-        // replacement method.
-        if (threadId === null && thread.length) {
-          thread.some(function(d, i) {
-            if (d.id === draft.id) {
-              this.delete(d);
-              stored = null;
-              return true;
-            }
-          }, this);
-        }
+      this.store();
 
-        thread.push(draft);
-        draftIndex.set(threadId, thread);
-        this.store();
-      }
       return this;
     },
+
     /**
-     * delete
-     *
-     * Delete a draft record from the collection.
-     *
-     * @param  {Draft} draft draft to delete.
-     *
-     * @return {Drafts} return the Drafts object.
+     * Deletes a draft record from Drafts in-memory collection.
+     * @param  {Draft} draft Draft instance to delete.
+     * @returns {Drafts} Returns the Drafts object.
      */
     delete: function(draft) {
-      var thread;
-      var index;
-
-      if (draft) {
-        thread = draftIndex.get(draft.threadId);
-        if (!thread) {
-          return this;
-        }
-        index = thread.indexOf(draft);
```

Comment:
not sure either; I think the story is that we used indexOf first and then we found it was not enough